tests/integrationv2/test_renegotiate.py

```diff
@@ -0,0 +1,259 @@
+import copy
+import os
+import pytest
+import time
+import random
+from collections import namedtuple
+
+from configuration import available_ports, ALL_TEST_CIPHERS, ALL_TEST_CURVES, MINIMAL_TEST_CERTS, PROTOCOLS
+from common import ProviderOptions, Protocols, Curves
+from fixtures import managed_process
+from providers import Provider, S2N, OpenSSL, GnuTLS
+from utils import invalid_test_parameters, get_parameter_name
+
+
+# TLS1.3 does not support renegotiation
+TEST_PROTCOLS = [x for x in PROTOCOLS if x.value < Protocols.TLS13.value]
+
+
+# Command line options to enable renegotiation
+S2N_RENEG_OPTION = "--renegotiation"
+S2N_RENEG_ACCEPT = "accept"
+S2N_RENEG_REJECT = "reject"
+
+
+# Output indicating renegotiation state
+S2N_RENEG_START_MARKER = "RENEGOTIATE"
+S2N_RENEG_SUCCESS_MARKER = "s2n is ready, again"
+OPENSSL_RENEG_REQ_MARKER = "SSL_accept:SSLv3/TLS write hello request"
+OPENSSL_RENEG_WARN_MARKER = "SSL3 alert read:warning:no renegotiation"
+
+
+# Methods to check renegotiation state
+def renegotiate_was_requested(openssl_results):
+    return to_bytes(OPENSSL_RENEG_REQ_MARKER) in openssl_results.stderr
+def renegotiate_was_rejected(openssl_results):
+    return to_bytes(OPENSSL_RENEG_WARN_MARKER) in openssl_results.stderr
+def renegotiate_was_started(s2n_results):
+    return to_bytes(S2N_RENEG_START_MARKER) in s2n_results.stdout
+def renegotiate_was_successful(s2n_results):
+    return renegotiate_was_started(s2n_results) and \
+           to_bytes(S2N_RENEG_SUCCESS_MARKER) in s2n_results.stdout
+
+
+# Basic conversion methods
+def to_bytes(val):
+    return str(val).encode('utf-8')
+def to_marker(val):
+    return bytes(val).decode('utf-8')
+
+
+"""
+The integrationv2 framework controls process IO with "markers".
+Messages can only be written to stdin if "send_marker"s are read from stdio.
+This results in the restriction that a process can only send in response to its own output,
+and also makes reasoning about message ordering very difficult.
+
+Because testing renegotiation requires multiple messages to be sent and received,
+we construct the "data_to_send", "send_marker" and "close_marker" lists required
+by the integrationv2 framework programmatically based on a "Msg" abstraction
+that is easier to visualize and reason about.
+
+For simplicity, "Msg" currently assumes:
+- s2n sends first. This lets us assume the first send_marker is S2N.get_send_marker().
```

Comment:
If the "r\n" in `RENEG_MESSAGES` is specific to openssl anyway, I wonder if it makes sense to just pass the provider into `Msg` rather than `mode`? That way you could just call message.provider.get_send_marker() rather than assume it's S2N. Not sure.

Reply:
provider.get_send_marker() is annoyingly unreliable for some providers. I use S2N.get_send_marker() wherever possible. In general I use s2n output over provider output wherever possible :)

Also, what if we someday do support server renegotiation and want to add an S2N to S2N integration test? Both providers would be the same. We really need mode.

Reply:
Ah okay, that makes sense!